core/jvm/src/SegmentPool.kt

```diff
@@ -83,55 +142,148 @@ internal actual object SegmentPool {
+        if (segment.copyTracker?.removeCopyIfShared() == true) return // This segment cannot be recycled.
 
-        val firstRef = firstRef()
+        segment.pos = 0
+        segment.owner = true
+        while (true) {
+            val firstRef = firstRef()
 
-        val first = firstRef.get()
-        if (first === LOCK) return // A take() is currently in progress.
-        val firstLimit = first?.limit ?: 0
-        if (firstLimit >= MAX_SIZE) return // Pool is full.
+            val first = firstRef.get()
+            if (first === LOCK) continue // A take() is currently in progress.
+            val firstLimit = first?.limit ?: 0
+            if (firstLimit >= MAX_SIZE) {
+                // L1 pool is full.
+                if (SECOND_LEVEL_POOL_TOTAL_SIZE > 0) {
+                    recycleL2(segment)
+                }
+                return
+            }
 
-        segment.next = first
+            segment.next = first
+            segment.limit = firstLimit + Segment.SIZE
+
+            if (firstRef.compareAndSet(first, segment)) {
+                return
+            }
+        }
+    }
+
+    @JvmStatic
+    private fun recycleL2(segment: Segment) {
         segment.pos = 0
-        segment.limit = firstLimit + Segment.SIZE
+        segment.owner = true
+
+        var bucket = l2BucketId()
+        var attempts = 0
+        var firstRef = hashBucketsL2[bucket]
+
+        while (true) {
+            val first = firstRef.get()
+            if (first === LOCK) continue // A take() is currently in progress.
+            val firstLimit = first?.limit ?: 0
+            if (firstLimit >= SECOND_LEVEL_POOL_BUCKET_SIZE) {
+                if (attempts < HASH_BUCKET_COUNT_L2) {
+                    attempts++
+                    bucket = (bucket + 1) and (HASH_BUCKET_COUNT_L2 - 1)
+                    firstRef = hashBucketsL2[bucket]
+                    continue
+                }
+                // L2 pool is full.
+                return
+            }
+
+            segment.next = first
+            segment.limit = firstLimit + Segment.SIZE
 
-        // If we lost a race with another operation, don't recycle this segment.
-        if (!firstRef.compareAndSet(first, segment)) {
-            segment.next = null // Don't leak a reference in the pool either!
+            if (firstRef.compareAndSet(first, segment)) {
+                return
+            }
         }
     }
 
+    @JvmStatic
+    actual fun tracker(): SegmentCopyTracker = RefCountingCopyTracker()
+
     private fun firstRef(): AtomicReference<Segment?> {
         // Get a value in [0..HASH_BUCKET_COUNT) based on the current thread.
         @Suppress("DEPRECATION") // TODO: switch to threadId after JDK19
-        val hashBucket = (Thread.currentThread().id and (HASH_BUCKET_COUNT - 1L)).toInt()
-        return hashBuckets[hashBucket]
+        val bucket = (Thread.currentThread().id and (HASH_BUCKET_COUNT - 1L)).toInt()
```

Comment:
Filed https://youtrack.jetbrains.com/issue/KT-69804/Introduce-Thread.threadId-that-encapsulates-either-Thread.getId-or-Thread.threadId and https://youtrack.jetbrains.com/issue/KT-69803/Support-multi-release-JAR-capabilities-in-the-standard-library